lib/src/boollist.dart

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:collection' show ListMixin;
+import 'dart:typed_data' show Uint32List;
+
+import 'unmodifiable_wrappers.dart' show NonGrowableListMixin;
+
+/// A BoolList implementation for Dart to store boolean values.
+///
+/// Uses list of integers as internal storage to reduce memory usage.
+///
+/// Internal storage of the growable [BoolList] is dynamically changed in the
+/// following cases:
+/// * when required length greater than preallocated space, storage expands its
+///   space to `2 * length` for further length expandings;
+/// * when required length less than the half of the current [length], storage
+///   shrinks availiable space.
+abstract class BoolList with ListMixin<bool> {
+  static const int _entryShift = 5;
+
+  static const int _bitsPerEntry = 32;
+
+  static const int _entrySignBitIndex = 31;
+
+  int _length;
+
+  Uint32List _data;
+
+  BoolList._(this._data, this._length);
+
+  factory BoolList._selectType(int length, bool growable) {
+    if (growable) {
+      return _GrowableBoolList(length);
+    } else {
+      return _NonGrowableBoolList(length);
+    }
+  }
+
+  /// Creates a [BoolList] with given length.
+  ///
+  /// The created list is fixed-length if [length] is provided.
+  /// Al initial values are `false`.
```

Comment:
I wouldn't copy the `List` constructor behavior. We removed it for a reason.

I'm slightly torn on whether a growable list is something you'll use often or not. That matters for which constructor should be the default.
I am *leaning* towards making the fixed-length constructor the default, which means that `BoolList` should take a length and default to a fixed-length false-filled list of that length.
Then we can add another constructor for growable lists.
```dart
  /// Creates a list of booleans with the provided length.
  ///
  /// The list is initially filled with the [fill] value, and
  /// the list is growable if [growable] is true.
  BoolList(int length, {bool fill = false, bool growable = false}) ...

  /// Creates an empty list of booleans.
  /// 
  /// The list defaults to being growable unless [growable] is `false`.
  /// If [capacity] is provided, and [growable] is not `false`,
  /// the implementation will attempt to make space for that
  /// many elements before needing to grow its internal storage.
  BoolList.empty({bool growable = true, int capacity = 0}) ...
```

That would replace the current `.filled` and unnamed constructor.

Reply:
Do we need to keep `.generate` and `.from` constructors?

Reply:
I'd keep them. They're obviously not *necessary*, you can always just create a list of the required length and fill in the data, but they are convenient in some situations. (Now, if we could make bool-list literals, then we probably wouldn't need those two, but we can't do `BoolList[for (var i = 0; i < 10; i++) generator(i)]` or `BoolList[...source]`. Yet!)